db/migrate/20250611083925_fix_stale_billing_entity_sequential_id_to_be_uniq.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+class FixStaleBillingEntitySequentialIdToBeUniq < ActiveRecord::Migration[8.0]
+  class Invoice < ApplicationRecord
+  end
+
+  class BillingEntity < ApplicationRecord
+  end
+
+  def up
+    # BillingEntities::ChangeInvoiceNumberingService -- if we switch from per_customer
+    # to per_billing_entity, we'll recalculate the billing_entity_sequential_id
+    BillingEntity.where(document_numbering: "per_customer").find_each do |billing_entity|
+      # find invoices with duplicated billing_entity_sequential_id
+      duplicates = Invoice.where(billing_entity_id: billing_entity.id)
+        .non_self_billed.with_generated_number
+        .where.not(billing_entity_sequential_id: nil)
+        .group(:billing_entity_sequential_id)
+        .having("COUNT(*) > 1")
+        .pluck(:billing_entity_sequential_id)
+      next if duplicates.empty?
+
+      # update the billing_entity_sequential_id to NULL for the duplicated invoices
+      # rubocop:disable Rails/SkipsModelValidations
+      Invoice.where(billing_entity_id: billing_entity.id)
+        .non_self_billed.with_generated_number
+        .where.not(billing_entity_sequential_id: nil)
+        .where(billing_entity_sequential_id: duplicates)
+        .update_all("billing_entity_sequential_id = NULL")
+      # rubocop:enable Rails/SkipsModelValidations
+    end
+
+    BillingEntity.where(document_numbering: "per_billing_entity").find_each do |billing_entity|
+      # group invoices by billing_entity_sequential_id and find groups with more than 1 invoice
+      duplicates = Invoice.where(billing_entity_id: billing_entity.id)
+        .non_self_billed.with_generated_number
+        .where.not(billing_entity_sequential_id: nil)
+        .group(:billing_entity_sequential_id)
+        .having("COUNT(*) > 1")
+        .pluck(:billing_entity_sequential_id)
+      next if duplicates.empty?
+
+      invoices_count = Invoice.where(billing_entity_id: billing_entity.id, billing_entity_sequential_id: duplicates).count
+      latest_invoice = Invoice.where(billing_entity_id: billing_entity.id, billing_entity_sequential_id: duplicates).order(:created_at).last
+      Rails.logger.info "Found #{duplicates.count} duplicates for billing_entity: #{billing_entity.name}; Affected invoices: #{invoices_count}; Latest invoice: (#{latest_invoice.created_at})"
```

Comment:
creation is perhaps not quite relevant as the invoice could have been created 1 month ago and finalized now... we assign the number on finalization ... perhaps I'm missing the intention behind using `created_at`.

Reply:
this output is only for debugging purposes, to see that these invoices were not created in 2025;
before 2025 we didn't have :finalized_at, and within the same billing_entity/organization the grace_period mostly will be the same for all invoices (except when it's defined on customer), so for majority of invoices, even if they were finalized in month after creation, we'll get the correct order and will see the estimate of when the duplicated organization_sequential_id was assigned